.github/workflows/main.yml

```diff
@@ -1,101 +1,77 @@
 name: CI
 on: [push, pull_request]
 
-permissions:
-  contents: read #  to fetch code (actions/checkout)
-
 jobs:
   test:
     name: Test
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        build: [stable, beta, nightly, macos, win32, win64, mingw]
         include:
           - build: stable
-            os: ubuntu-latest
+            os: ubuntu-24.04
             rust: stable
           - build: beta
-            os: ubuntu-latest
+            os: ubuntu-24.04
             rust: beta
           - build: nightly
-            os: ubuntu-latest
+            os: ubuntu-24.04
             rust: nightly
           - build: macos
-            os: macos-latest
+            os: macos-14
             rust: stable
           - build: win32
-            os: windows-latest
+            os: windows-2022
             rust: stable-i686-pc-windows-msvc
           - build: win64
-            os: windows-latest
+            os: windows-2022
             rust: stable-x86_64-pc-windows-msvc
           - build: mingw
-            os: windows-latest
+            os: windows-2022
             rust: stable-x86_64-pc-windows-gnu
     steps:
-    - uses: actions/checkout@master
-    - name: Install Rust
-      run: |
-        rustup update ${{ matrix.rust }} --no-self-update
-        rustup default ${{ matrix.rust }}
-        cargo install cargo-hack
+    - uses: actions/checkout@v4
+    - name: Install toolchain
+      uses: dtolnay/rust-toolchain@master
+      with:
+        toolchain: ${{ matrix.rust }}
+    - uses: taiki-e/install-action@cargo-hack
```

Comment:
I'd rather not depend on any external actions in CI, and instead run the scripts ourselves. Doing that makes our CI less opaque, reduces our dependence on outside infrastructure, and limits our tie in to GitHub Actions itself.